feat(manager/terragrunt): support lockFileMaintenance

spilliams opened the pull request:
terragrunt configurations sometimes have no tf files, but do have terraform lockfiles.

<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

<!-- Describe what behavior is changed by this PR. -->

Adds a function `updateArtifacts` to the Terragrunt manager that wraps the Terraform manager's implementation. As requested in [this discussion](https://github.com/renovatebot/renovate/issues/13393#issuecomment-1007957758), it only calls the downstream function if the update type is `lockFileMaintenance`.

## Context

<!-- Describe why you're making these changes if it's not already explained in a corresponding issue. -->
<!-- If you're closing an existing issue with this pull request, use the keyword Closes #issue_number. -->
<!-- If you're referencing an issue with this pull request, put it in a Markdown list like this: - #issue_number. -->

The current Renovate behavior is that it normally does not update a Terraform lockfile if it is in a directory with no `.tf` files. This can be the case when the configuration is managed by Terragrunt, where the `terragrunt.hcl` includes a `terraform` block to call out to an external module.

The expected behavior is that when Renovate runs on a minimal Terragrunt configuration, it will update the versions and hashes stored in the `.terraform.lock.hcl`. This is a listing of what a minimal Terragrunt configuration contains:

```text
my-service
├── .terraform.lock.hcl
├── .terragrunt-cache/
└── terragrunt.hcl
```

and a minimal `terragrunt.hcl` is:

```hcl
terraform {
  source = "path/to/module"
}
```

Closes #13393

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

My test repository is [here](https://github.com/spilliams/terragrunt-renovate-test). The "before" PR was [this one](https://github.com/spilliams/terragrunt-renovate-test/pull/3) and the "after" PR was [this one](https://github.com/spilliams/terragrunt-renovate-test/pull/22).

I've run `yarn test` locally and it shows that coverage is unchanged from the before this branch:

```text
=============================== Coverage summary ===============================
Statements   : 99.99% ( 33708/33709 )
Branches     : 98.08% ( 10843/11055 )
Functions    : 100% ( 4076/4076 )
Lines        : 99.99% ( 33217/33218 )
================================================================================
```

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

viceice commented on lib/modules/manager/terragrunt/lockfile.ts:
```diff
@@ -0,0 +1,12 @@
+export async function updateArtifacts(
+  artifact: UpdateArtifact
+): Promise<UpdateArtifactsResult[] | null> {
+  if (artifact.config.updateType !== 'lockFileMaintenance') {
```
why only lockfile maintenance?

spilliams replied:
I was following the advice of @secustor in [this comment](https://github.com/renovatebot/renovate/issues/13393#issuecomment-1007957758).

Are there other update types that this wrapper should accept?

viceice replied:
don't you also need to update the lockfile when the provider version changes?

with current implementation the lockfile would only be updated on lockfile maintenance.

viceice replied:
@secustor do you know more on terragrunt tooling?

secustor replied:
I have a rough idea regarding it, but no details. I will look into it.

spilliams replied:
I think we want renovate to update the lockfile any time it would do so if the directory were a normal terraform configuration. As far as I can tell, that means there is no restriction on `updateType`, and this code becomes a lot simpler.

viceice replied:
terraform manager only updates lockfiles if provider version changes or lockfile maintenance

spilliams replied:
I've updated this branch to remove the wrapper function, and to instead import the whole terraform `updateArtifacts`. I'll update the PR description shortly, but I've run the tests and code coverage is maintained at pre-branch levels.

secustor replied:
Lockfile updates on dependency updates are currently not possible without some bigger additions to the code.    

Artifact update work currently with a relation from the package file to lock files.     
In case of Terraform this is from a tf file with a reference to  a `.terraform.lock.hcl` in the same folder.    
For Terragrunt these files are `terragrunt.hcl` and `.terraform.lock.hcl`, which have to be again in the same folder.

Therefore there has to be an update in `terragrunt.hcl` to trigger an update in `.terraform.lock.hcl`. `lockfileMaintenance` is an exception for this as here only the relation has to be detected.    
If this has happened, the constraints are used to find a fitting version. 

Therefore you have to ensure that only updateType with `lockfileMaintenance` as value leads to a execution of the `updateArtifacts` function of the `terraform` manager.

spilliams replied:
I think I understand this so far. Let me ask something new to test my understanding.

Say I have a `terragrunt.hcl` with the following contents:

```
terraform {
  source = "git@github.com:user/repo.git//path/to/module?ref=1.0.0"
}
```

This module at version 1.0.0 requires the `aws` provider version `3.x`.

Now let's say that module releases a new version `2.0.0`, which updates its required provider version to `4.x`.

When Renovate runs to upgrade my Terragrunt configuration to the latest version of the module, that upgrade must include a change to the lockfile artifact.

In this scenario, did Renovate run with an updateType of `lockfileMaintenance`?

secustor replied:
> In this scenario, did Renovate run with an updateType of lockfileMaintenance?

No, in this specific case it will run with `major` as `updateType`, because Renovate updates the module from `1.0.0` to `2.0.0`. It has no knowledge what is in the module, it only looks at the version it has currently updated. 

> [perhaps this question is evidence that I should be building a fixture and integration test...]

A good indicator is that whenever you change the behaviour it should be reflected in some tests.

spilliams replied:
> It has no knowledge what is in the module, it only looks at the version it has currently updated.

So in this case, a `major` update would require further (manual) intervention before it could be merged, which seems like reasonable behavior.

If we change the scenario slightly my point might be better Illustrated. Say that the module is upgrading from version `1.0.0` to `1.1.0`, and its underlying provider requirements are changing from `3.76.0` to `3.76.1`. From Renovate's perspective, this is a `minor` updateType, and only affects the module version in `terragrunt.hcl`, it does not make changes to the lockfile. This scenario _also_ requires manual intervention before it can be reliably deployed, which I think is not reasonable behavior. I would expect anyone doing a "routine upgrade" to a module (such as a minor version upgrade) to update any lockfile artifacts as necessary. Does this go against Renovate policy or ideology?

spilliams replied:
I've performed a new manual test. I ran Renovate as a `minor` update to both a terraform configuration and a terragrunt configuration. I [first did this](https://github.com/spilliams/terragrunt-renovate-test/pull/14) from the `main` branch of my fork of the repo, and [then from this feature branch](https://github.com/spilliams/terragrunt-renovate-test/pull/15). The behavior was the same in both cases: Renovate changed the version number of the module, but did nothing to the associated lockfile.

At this point I don't understand why a wrapper function with a condition for `config.updateType === 'lockFileMaintenance'` is necessary.  I would expect the Terragrunt manager to behave very much like the Terraform manager in all cases, and indeed my testing shows that it does behave the same.

viceice reviewed (changes requested):
needs unit test

spilliams commented:
> needs unit test

@viceice What kind of unit test were you thinking? The logic in this function is straightforward enough to me that I didn't think it warrants one, but I'm also new to this codebase.

spilliams commented:
When I run `yarn test` from my fork's main branch (up to date with upstream remote), I get 2 failed tests and less than 100% coverage:

```text
=============================== Coverage summary ===============================
Statements   : 99.99% ( 33695/33696 )
Branches     : 98.08% ( 10838/11050 )
Functions    : 100% ( 4074/4074 )
Lines        : 99.99% ( 33205/33206 )
================================================================================
Jest: "global" coverage threshold for statements (100%) not met: 99.99%
Jest: "global" coverage threshold for lines (100%) not met: 99.99%

Summary of all failing tests
 FAIL  lib/util/github/graphql/cache-strategies/package-cache-strategy.spec.ts (170 MB heap size)
  ● util/github/graphql/cache-strategies/package-cache-strategy › reconciles old cache record with new items

    expect(received).toEqual(expected) // deep equality

    - Expected  - 1
    + Received  + 1

    @@ -22,8 +22,8 @@
                "version": "4",
              },
            },
            "updatedAt": "2022-10-30T12:00:00.000-07:00",
          },
    -     21600,
    +     21660,
        ],
      ]

      46 |     expect(res).toEqual([...Object.values(oldItems), newItem]);
      47 |     expect(isPaginationDone).toBe(false);
    > 48 |     expect(cacheSet.mock.calls).toEqual([
         |                                 ^
      49 |       [
      50 |         'foo',
      51 |         'bar',

      at Object.<anonymous> (lib/util/github/graphql/cache-strategies/package-cache-strategy.spec.ts:48:33)

 FAIL  lib/modules/platform/github/index.spec.ts (22.313 s, 550 MB heap size)
  ● modules/platform/github/index › getBranchPr(branchName) › aborts reopen if PR is too old

    Missing mocks!
     *   POST https://api.github.com/repos/some/repo/git/refs

      51 |   missingLog = [];
      52 |   if (missing.length && throwOnPending) {
    > 53 |     throw new Error(`Missing mocks!\n * ${missing.join('\n * ')}`);
         |           ^
      54 |   }
      55 |   if (!isDone && throwOnPending) {
      56 |     throw new Error(`Pending mocks!\n * ${pending.join('\n * ')}`);

      at clear (test/http-mock.ts:53:11)
      at Object.<anonymous> (test/http-mock.ts:120:3)


Test Suites: 2 failed, 590 passed, 592 total
Tests:       2 failed, 11905 passed, 11907 total
Snapshots:   864 passed, 864 total
Time:        233.747 s
Ran all test suites.
```

When I run `yarn test` from the development branch, I get the same two failures and a slightly worse coverage:

```text
=============================== Coverage summary ===============================
Statements   : 99.98% ( 33700/33704 )
Branches     : 98.07% ( 10838/11051 )
Functions    : 99.97% ( 4075/4076 )
Lines        : 99.98% ( 33209/33213 )
================================================================================
```

I'll update my branch to try and at least get the Functions coverage back to 100%

viceice reviewed (commented):
did you test the latest version against real repos? I'd like to see some working sample PR for different update types.

viceice commented:
> > did you test the latest version against real repos?
> 
> @viceice Yes, I tested on my public repository [here](https://github.com/spilliams/terragrunt-renovate-test). I've summarized all tests to date in [this wiki page](https://github.com/spilliams/terragrunt-renovate-test/wiki/TestCases)

Idon't see a pr open with a changed lockfile in your test rep 👀

spilliams commented:
> Idon't see a pr open with a changed lockfile in your test rep 👀

That's because I close the PRs as I go, to provide a clean testing workbench (when I run `yarn start ...` it might create multiple PRs at once). See all PRs in that repo [here](https://github.com/spilliams/terragrunt-renovate-test/pulls?q=is%3Apr).

spilliams commented:
I just ran it again, from commit [5fa37874f](https://github.com/spilliams/renovate/commit/5fa37874fadce91e462ad78f95ef705611e923aa), and left the resulting PRs open for visibility ([required_provider minor](https://github.com/spilliams/terragrunt-renovate-test/pull/16) and [lockfile](https://github.com/spilliams/terragrunt-renovate-test/pull/17)).

viceice commented:
as you can see, the provider PR is missing a lockfile update. so it needs more things to do here to support that case.

maybe we simply need to match the extract properties to terraform.

@secustor can you help?

secustor commented:
🤔  for this to work we need to follow the reference to the modules and then do a full extraction with the terraform manager. This will give us the updated dependencies which we than can use to update the lock file. 

The problem with that, besides that it is costly, is that we have to leave the directory where the package file is located.    
We forbid this currently in most cases because of security concerns. Further we would have to inject updates of other managers ( `terraform` in this case ) to `updateArtifacts`. 
@rarkins @viceice   WDYT regarding this?

rarkins commented:
> The problem with that, besides that it is costly, is that we have to leave the directory where the package file is located. We forbid this currently in most cases because of security concerns. 

Leave it to where? It's ok to traverse the repo itself

> Further we would have to inject updates of other managers ( `terraform` in this case ) to `updateArtifacts`. @rarkins @viceice WDYT regarding this?

What type of updates? Do you mean run multiple binaries?

secustor commented:
> Leave it to where? It's ok to traverse the repo itself
No still inside the repo but I don't think we have function to do find file patterns safely. We have one if we know the file names tough. 

> What type of updates? Do you mean run multiple binaries?
The resulting updates from the `terraform` manager before hitting the `updateArtifacts` phase.

secustor commented:
You need unit tests which call the `updateArtifacts` of the `terragrunt` manager with `updateType` set to `lockfileMaintenance`. If the `updateType` is set to something different, currently `updateArtifact` should return null.

At best you implement `updateArtifacts` in `terragrunt`, check for `updateType` and then return `null` or the result of the terraform manager.
   You can take a look how the `terraform` manager tests `lockfileMaintenance` here:  https://github.com/renovatebot/renovate/blob/39493b715ba06881111fb6df65766905a51f78a1/lib/modules/manager/terraform/lockfile/index.spec.ts#L317

relsqui commented:
Hi, I'm a teammate of @spilliams's taking a look at this while he's focused on other work. Am I understanding these requested changes correctly?
* rather than just re-exporting the Terraform function from the Terragrunt manager, put it in a small wrapper which returns `null` if `updateType` isn't `lockFileMaintenance` (since we're limiting the scope to just that update type, and the Terraform function still does other things in that case)
* add a test that verifies that the wrapper returns null when `updateType` isn't `lockFileMaintenance`
* add a test that verifies the wrapper does the same thing as the Terraform lock file updater when the `updateType` is `lockFileMaintenance`. I read through the existing test you linked and, while my Jest is pretty rusty, I didn't notice anything that would be different if we reimplemented it for the Terragrunt manager. is it still useful to copy that code over so it can run from the Terragrunt context as well?

spilliams commented:
I've reverted to a version of this branch that has a wrapper function with conditional logic, and a unit test ensuring the quality of that conditional logic. I ran a test on my public test repo with the latest commit on this branch, and it shows the [desired behavior](https://github.com/spilliams/terragrunt-renovate-test/pull/22/files).

secustor commented on lib/modules/manager/terragrunt/lockfile.ts:
```diff
@@ -0,0 +1,12 @@
+  artifact: UpdateArtifact
+): Promise<UpdateArtifactsResult[] | null> {
+  if (artifact.config.updateType !== 'lockFileMaintenance') {
+    return null;
```
```suggestion
    logger.debug(`UpdateType ${artifact.config.updateType} is not supported for terragrunt`)
    return null;
```

spilliams commented:
(sorry about the review request spam, I didn't realize it would only have one request out at a time)

viceice reviewed (commented):
lint error

rarkins reviewed (changes requested):
Clarify on lock file maintenance vs all other type of lock file updates

viceice commented on lib/modules/manager/terragrunt/index.ts:
```diff
@@ -2,6 +2,7 @@ import { GitTagsDatasource } from '../../datasource/git-tags';
 import { GithubTagsDatasource } from '../../datasource/github-tags';
 import { TerraformModuleDatasource } from '../../datasource/terraform-module';
 
+export { updateArtifacts } from './lockfile';
```
please rename, because all other managers are using `artifacts`
```suggestion
export { updateArtifacts } from './artifacts';
```


we should rename terraform manager too in a separate PR.

spilliams commented:
@rarkins Thank you for taking a look last week! I've updated the PR description, and added some simple documentation to the Terragrunt manager readme (which certainly could be insufficient, so I don't mind improving it if you have more recommendations).

spilliams commented:
@rarkins are there other changes you'd like me to make to the documentation?

rarkins reviewed (changes requested):
I think it would be good to document the limitations which exist currently (e.g. the only way a lock file is updated is through lock file maintenance?).

rarkins commented on lib/modules/manager/terragrunt/readme.md:
```diff
@@ -20,3 +20,7 @@ terraform {
+
+### Terraform lockfiles
+
+The Terragrunt manager supports [lock file maintenance](https://docs.renovatebot.com/configuration-options/#lockfilemaintenance) for `.terraform.lock.hcl` artifacts. These artifacts will be updated if and only if the update type is `lockFileMaintenance`.
```
```suggestion
The Terragrunt manager supports [lock file maintenance](https://docs.renovatebot.com/configuration-options/#lockfilemaintenance) for `.terraform.lock.hcl` artifacts.
These artifacts will be updated if and only if the update type is `lockFileMaintenance`.
```

viceice commented on lib/modules/manager/terragrunt/artifacts.spec.ts:
```diff
@@ -0,0 +1,74 @@
+      'bump',
+    ];
+
+    for (const t of updateTypes) {
```
use `it.each` to have separate tests.

spilliams replied:
Can do! I didn't realize that feature exists but it makes sense that it does.

viceice commented on lib/modules/manager/terragrunt/readme.md:
```diff
@@ -1,3 +1,5 @@
+Extracts dependencies from `terragrunt.hcl` files, and the associated `.terraform.lock.hcl` files too.
```
this seems wrong, as we don't extract the lock files?

spilliams replied:
Ah, thanks for the correction. I'm still learning what the terms and processes are in this project.

rarkins commented:
If a terragrunt file has a lock file, shouldn't the lock file be updated any time the package file is updated?

If I understand this PR, it would only update this lock file independently on a lock file maintenance schedule, while all other PRs would have an un-updated and hence out of date lock file?

viceice commented on lib/modules/manager/terragrunt/artifacts.spec.ts:
```diff
@@ -0,0 +1,75 @@
+  ];
+
+  beforeEach(() => {
+    jest.resetAllMocks();
+    jest.resetModules();
```
```suggestion
```

the first does nothing without dynamic imports
the second does nothing, because you don't set any explicit mocks.

The mocks are cleared globally here:
https://github.com/renovatebot/renovate/blob/85fbad1029d4a10ff07b93152324b68b8a9fd990/jest.config.ts#L42

viceice commented on lib/modules/manager/terragrunt/artifacts.spec.ts:
```diff
@@ -0,0 +1,75 @@
+import { mocked } from '../../../../test/util';
+import { GlobalConfig } from '../../../config/global';
+import type { UpdateType } from '../../../config/types';
+import * as TerraformLockfile from '../terraform/lockfile';
```
```suggestion
import * as terraformLockfile from '../terraform/lockfile';
```

viceice commented on lib/modules/manager/terragrunt/artifacts.spec.ts:
```diff
@@ -0,0 +1,75 @@
+import { updateArtifacts } from './artifacts';
+
+jest.mock('../terraform/lockfile');
+const mockTFUpdateArtifacts = mocked(TerraformLockfile).updateArtifacts;
```
```suggestion
```

viceice commented on lib/modules/manager/terragrunt/artifacts.spec.ts:
```diff
@@ -0,0 +1,75 @@
+      newPackageFileContent: '',
+      config: localConfig,
+    });
+    expect(mockTFUpdateArtifacts.mock.calls).toBeArrayOfSize(1);
```
```suggestion
    expect(terraformLockfile.updateArtifacts).toHaveBeenCalledOnce();
```

viceice commented on lib/modules/manager/terragrunt/artifacts.spec.ts:
```diff
@@ -0,0 +1,75 @@
+        newPackageFileContent: '',
+        config: localConfig,
+      });
+      expect(mockTFUpdateArtifacts.mock.calls).toBeArrayOfSize(0);
```
```suggestion
      expect(terraformLockfile.updateArtifacts).not.toHaveBeenCalled();
```

rarkins commented:
> If a terragrunt file has a lock file, shouldn't the lock file be updated any time the package file is updated?
> 
> 
> 
> If I understand this PR, it would only update this lock file independently on a lock file maintenance schedule, while all other PRs would have an un-updated and hence out of date lock file?

This remains unanswered?

spilliams commented:
Thanks for pinging that @rarkins. Here are my thoughts, but I believe @secustor has some thoughts as well:

> If a terragrunt file has a lock file, shouldn't the lock file be updated any time the package file is updated?

Terragrunt (and Terraform) deployments don't always have package files. There may be updated constraints on dependency versions, but in general for my use cases I keep my coded dependency constraints as low as possible and try to keep my lockfile updated with the latest versions of my dependencies. This usually means running `terraform init -upgrade` locally.

> If I understand this PR, it would only update this lock file independently on a lock file maintenance schedule, while all other PRs would have an un-updated and hence out of date lock file?

Perhaps. I'm not really sure since I'm so new to the inner workings of Renovate. I'll say two things here:

1. As a Terraform/Terragrunt user, I expect any tool that works on Terraform to behave _identically_ with a Terragrunt implementation. This is why I [proposed early in this PR](https://github.com/renovatebot/renovate/pull/20833/commits/ee132c911a97c6e33ae5320284155886df3be9a3) that the terragrunt manager wholly import the terraform manager's `updateArtifacts` function.
2. Sebastian had recommended both [in this PR](https://github.com/renovatebot/renovate/pull/20833#issuecomment-1486512668) and in the [prior discussion](https://github.com/renovatebot/renovate/issues/13393#issuecomment-1007957758) that we use a wrapper function that only updates lockfiles under certain conditions. I admit I don't fully understand the logic behind this recommendation, but I deferred to their knowledge of the system and implemented the solution as requested.

rarkins commented:
@secustor can you clarify why you think only lock file maintenance is necessary?

secustor commented:
I think it is necessary, but requires larger changes in the code base to work efficiently outside of `lockFileMaintenance` as we have to rerun the terraform manager for specific directories. This could be solved by caching based on the directories. 

I see `lockFileMaintenance` as good first step to get basic support out and gauge if there is an interest to extend the manager.

relsqui commented:
Thank you maintainers! We just confirmed this works on our huge Terragrunt repo, and it's going to save us so much time.

norman-zon commented:
@ secustor I would like to express my interest to have the functionality implemented in the terraform manager.
My planned workflow looks like this:

1) When adding a new `terragrunt.hcl` to a repo I run `terragrunt init -upgrade` locally once.
2) renovate runs provider updates on a infrequent schedule (once a month would be fine) and also updates the lockfiles along the providers
3) The renovate PRs are tested with Atlantis
4) If Atlantis shows no diffs I merge the PRs

The current implementation fails at point 3. because Atlantis is run with `terragrunt init` (I would like to avoid to run it with `init upgrade` in case someone accidentally removes the version pinning). 

So having the lockfile updates integrated with the provider upgrades would be the preferred and IMHO most clean solution here.
Otherwise I would have to merge the renovate PRs without being able to have them automatically tested and will see breakage only after the seperate `lockFileMaintenance` is run.

secustor commented:
@norman-zon  Please create a discussion, rather than a comment for feature requests.
